### ISSUE TITLE
Stopping criteria other than maxeval in Python

### ISSUE THREAD

**Issue by Koris** [stevengj/nlopt#59]
I am having trouble using stopping criteria other than maxeval in python. If I use a criterium like ftol_rel, nlopt does not seem to take the stopping criterium into account. However, the problem seems to depend on the formulation of the optimization problem. Of course, the test problem on http://ab-initio.mit.edu/wiki/index.php/NLopt_Tutorial#Example_in_Python works with any of the stopping criteria (test code also on http://pastebin.com/aJZznFzb).

However, if I apply it on a different problem
min f(x)
with 0 <= x <= 1

The stopping criteria 'do not work': to me, the ftol_rel is reached much earlier in the iteration. A MWE is on http://pastebin.com/wpgA1bbN
Any thoughts?

**Comment by stevengj** [comment]
Note that, in MMA, the ftol tolerance is [only applied](https://github.com/stevengj/nlopt/blob/5af26227f380e45482fb8ce076d91c3b4c7d2125/mma/mma.c#L369-L370) to so-called "outer" iterations in the algorithm, ones which satisfy both the constraints and the "conservativity" of the convex approximants.

**Comment by Koris** [comment]
I have taken a look at this issue again. In the mean time, I've read Svanberg's 2002 paper.

There's something that doesn't seem to match in the MWE posted above. As you say, it is probably stuck in the inner iterations. However, the paper of Svanberg states in the introduction "Each new inner iteration requires function values, but no derivatives, calculated at the optimal solution of the most recent subproblem." I check for this in Python using grad.size and it always evaluates the gradient.
Also, the constraint value is negative after the first two or three iterations, so I don't 'see' why it should be in the inner iterations.
Can you shed light on this issue?

**Comment by deepauck** [comment]
Hi Koris,

Did you figure out a solution to the issue stated above? I too am stuck here.

Best,
Deepak

**Comment by Koris** [comment]
No.. In my opinion the MMA needs a rewrite, but that's beyond my capabilities.

**Comment by deepauck** [comment]
Thanks Koris, I too got the same impression since I had been struggling with this issue for quite sometime. I also came across issues that MMA gets stuck sometimes to a non optimal solution. 

However, do you know of a way, where I can manually (if-condition based) terminate MMA based on checking some condition?
